include/gsl.h

```diff
@@ -122,15 +122,15 @@ class not_null
     not_null<T>& operator=(std::nullptr_t) = delete;
 	not_null<T>& operator=(int) = delete;
     
-    T get() const {
-#ifdef _MSC_VER
-        __assume(ptr_ != nullptr);
-#endif
-        return ptr_;
-    } // the assume() should help the optimizer
-
-    operator T() const {  return get(); }
-    T operator->() const { return get(); }
+    // if T has .get() forward its value wrapped in not_null
+    template<typename U = T, typename = decltype(std::declval<U>().get())>
```

Comment:
nit: most places in this code doing enable_if type things use the name Dummy for the template parameter